lib/galaxy/jobs/command_factory.py

```diff
@@ -109,12 +109,13 @@ def build_command(
     if container_monitor_command:
         commands_builder.prepend_command(container_monitor_command)
 
+    working_directory = remote_job_directory or job_wrapper.working_directory
+    commands_builder.capture_return_code(join(working_directory, f"galaxy_{job_wrapper.job_id}.ec"))
     if include_work_dir_outputs:
         __handle_work_dir_outputs(commands_builder, job_wrapper, runner, remote_command_params)
 
     if stdout_file and stderr_file:
         commands_builder.capture_stdout_stderr(stdout_file, stderr_file)
-    commands_builder.capture_return_code()
```

Comment:
Slightly concerned that the order of these 2 has been reversed, is `capture_stdout_stderr()` now catching the right output? I don't actually understand the Bash code in that method, but maybe a unit test with `stdout_file` and `stderr_file` could help prevent regressions.

Reply:
Either position should be fine. This was the idea: https://stackoverflow.com/questions/692000/how-do-i-write-standard-error-to-a-file-while-using-tee-with-a-pipe/692407#692407, and https://github.com/galaxyproject/galaxy/pull/6958#issuecomment-466087700 contains an example of a job script

Reply:
But now that I look at it more closely, I don't think it does what I wanted it to do, which is to print to stdout/stderr and to stdout and stderr files, so that job pods could stream the stdout/stderr.
For that I think the right pattern would have been
```diff
diff --git a/lib/galaxy/jobs/command_factory.py b/lib/galaxy/jobs/command_factory.py
index a30ea58338..989ae5f4e4 100644
--- a/lib/galaxy/jobs/command_factory.py
+++ b/lib/galaxy/jobs/command_factory.py
@@ -282,14 +282,14 @@ class CommandsBuilder:
         self.append_command("; ".join(c for c in commands if c))
 
     def capture_stdout_stderr(self, stdout_file, stderr_file):
-        self.prepend_command("""out="${TMPDIR:-/tmp}/out.$$" err="${TMPDIR:-/tmp}/err.$$"
+        self.prepend_command(
+            f"""out="${{TMPDIR:-/tmp}}/out.$$" err="${{TMPDIR:-/tmp}}/err.$$"
 mkfifo "$out" "$err"
 trap 'rm "$out" "$err"' EXIT
-tee -a stdout.log < "$out" &
-tee -a stderr.log < "$err" >&2 &""",
-                             sep="")
-        self.append_command(f"> '{stdout_file}' 2> '{stderr_file}'",
-                            sep="")
+tee -a '{stdout_file}' < "$out" &
+tee -a '{stderr_file}' < "$err" >&2 &""",
+        sep="")
+        self.append_command(f'> "$out" 2> "$err"', sep="")
 
     def capture_return_code(self, exit_code_path):
         self.append_command(CAPTURE_RETURN_CODE)

```

Reply:
Thanks, that makes much more sense!